secretservice/secretservice.go

```diff
@@ -0,0 +1,305 @@
+package secretservice
+
+import (
+	"fmt"
+	"math/big"
+	"time"
+
+	dbus "github.com/guelfey/go.dbus"
+	errors "github.com/pkg/errors"
+)
+
+const SecretServiceInterface = "org.freedesktop.secrets"
+const SecretServiceObjectPath dbus.ObjectPath = "/org/freedesktop/secrets"
+const DefaultCollection dbus.ObjectPath = "/org/freedesktop/secrets/aliases/default"
+
+type authenticationMode string
+
+const AuthenticationPlain authenticationMode = "plain"
+const AuthenticationDHIETF1024SHA256AES128CBCPKCS7 authenticationMode = "dh-ietf1024-sha256-aes128-cbc-pkcs7"
+
+const NilFlags = 0
+
+type Attributes map[string]string
+
+type Secret struct {
+	Session     dbus.ObjectPath
+	Parameters  []byte
+	Value       []byte
+	ContentType string
+}
+
+type PromptCompletedResult struct {
+	Dismissed bool
+	Paths     dbus.Variant
+}
+
+type SecretService struct {
+	conn     *dbus.Conn
+	signalCh <-chan *dbus.Signal
+}
+
+type Session struct {
+	Mode    authenticationMode
+	Path    dbus.ObjectPath
+	Public  *big.Int
+	Private *big.Int
+	AESKey  []byte
+}
+
+func NewService() (*SecretService, error) {
+	conn, err := dbus.SessionBus()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to open dbus connection")
+	}
+	signalCh := make(chan *dbus.Signal, 16)
+	conn.Signal(signalCh)
+	return &SecretService{conn: conn, signalCh: signalCh}, nil
+}
+
+func (s *SecretService) ServiceObj() *dbus.Object {
+	return s.conn.Object(SecretServiceInterface, SecretServiceObjectPath)
+}
+
+func (s *SecretService) Obj(path dbus.ObjectPath) *dbus.Object {
+	return s.conn.Object(SecretServiceInterface, path)
+}
+
+func (s *SecretService) OpenSession(mode authenticationMode) (session *Session, err error) {
+	var sessionAlgorithmInput dbus.Variant
+
+	session = new(Session)
+
+	session.Mode = mode
+
+	switch mode {
+	case AuthenticationPlain:
+		sessionAlgorithmInput = dbus.MakeVariant("")
```

Comment:
What does AuthenticationPlain mean?  Is the secret stored in plaintext?  If so, can we not support that mode?

Reply:
The secret is encrypted with the user's keyring password (configured whenever the user set up their keyring). This is just for encrypting the communication between keybase and the keyring, since it's possible something might be logging dbus message history. That's why it doesn't need to be authenticated. I can still remove it if you think it's a good idea though.

Reply:
That's fine, just making sure we weren't supporting a plaintext storage system.